src/geom/SpanSet.cc

```diff
@@ -0,0 +1,828 @@
+ public:
+            bool operator()(PrimaryRun const& pr, int yval) {
+                return pr.y < yval;
+            }
+            bool operator()(int yval, PrimaryRun const& pr) {
+                return yval < pr.y;
+            }
+};
+
+class ComparePrimaryRunM{
+ public:
+            bool operator()(PrimaryRun const& pr, int mval) {
+                return pr.m < mval;
+            }
+            bool operator()(int mval, PrimaryRun const& pr) {
+                return mval < pr.m;
+            }
+};
+}  // namespace
+
+namespace lsst { namespace afw {
+
+// Expose properties of the underlying vector containing spans such that the
+// SpanSet can be considered a container
+geom::SpanSet::const_iterator geom::SpanSet::begin() const {
+    // Return the constant version as SpanSets should be immutable
+    return _spanVector.cbegin();
+}
+
+geom::SpanSet::const_iterator geom::SpanSet::end() const {
+    // Return the constant version as SpanSets should be immutable
+    return _spanVector.cend();
+}
+
+geom::SpanSet::const_iterator geom::SpanSet::cbegin() const {
+    return _spanVector.cbegin();
+}
+
+geom::SpanSet::const_iterator geom::SpanSet::cend() const {
+    return _spanVector.cend();
+}
+
+geom::SpanSet::const_reference geom::SpanSet::front() const {
+    // Return constant version as SpanSets should be immutable
+    return const_cast<geom::Span &>(_spanVector.front());
+}
+
+geom::SpanSet::const_reference geom::SpanSet::back() const {
+    // Return constant version as SpanSets should be immutable
+    return const_cast<geom::Span &>(_spanVector.back());
+}
+
+geom::SpanSet::size_type geom::SpanSet::size() const {
+    return _spanVector.size();
+}
+
+bool geom::SpanSet::empty() const {
+    return _spanVector.empty();
+}
+
+// Default constructor, creates a null SpanSet which may be useful for
+// comparisons
+geom::SpanSet::SpanSet():_spanVector(), _bBox(), _area(0) {}
+
+// Construct a SpanSet from a Box2I object
+geom::SpanSet::SpanSet(geom::Box2I const & box):_bBox(box), _area(box.getArea()) {
+    int beginY = box.getMinY();
+    int endY = box.getMaxY();
+
+    int beginX = box.getMinX();
+    int endX = box.getMaxX();
+
+    for (int i = beginY; i < (endY+1); ++i) {
+        _spanVector.push_back(Span(i, beginX, endX));
+    }
+}
+
+// Construct a SpanSet from a std vector by copying
+geom::SpanSet::SpanSet(const std::vector<Span> & vec, bool normalize): _spanVector(vec) {
+    // If the incoming vector is zero, should create an empty spanSet
+    if (_spanVector.size() == 0) {
+        _bBox = geom::Box2I();
+        _area = 0;
+    } else {
+        if (normalize) {
+            _runNormalize();
+        }
+        _initialize();
+    }
+}
+
+// Construct a SpanSet from a std vector by moving
+geom::SpanSet::SpanSet(std::vector<Span> && vec, bool normalize): _spanVector(std::move(vec)) {
+    // If the incoming vector is zero, should create an empty SpanSet
+    if (_spanVector.size() == 0) {
```

Comment:
I am slightly embarrassed at how poor my c++ was when I started this project for things like this... I had no idea of basic std apis etc